cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -114,56 +135,93 @@ func createMCMManagerInternal(discoveryOpts cloudprovider.NodeGroupDiscoveryOpti
+	controlMachineClient := controlMachineClientBuilder.ClientOrDie("control-machine-client").MachineV1alpha1()
+	machineSharedInformers := controlMachineInformerFactory.Machine().V1alpha1()
+	azureMachineClassInformer := machineSharedInformers.AWSMachineClasses().Informer()
+	awsMachineClassInformer := machineSharedInformers.AWSMachineClasses().Informer()
+	machineInformer := machineSharedInformers.Machines().Informer()
+	machineSetInformer := machineSharedInformers.MachineSets().Informer()
+	machineDeploymentInformer := machineSharedInformers.MachineDeployments().Informer()
+	machineClassInformer := machineSharedInformers.MachineClasses().Informer()
+
+	// targetKubeconfig for the cluster for which nodes will be managed
+	targetKubeconfigPath := os.Getenv("TARGET_KUBECONFIG")
+	targetKubeconfig, err := clientcmd.BuildConfigFromFlags("", targetKubeconfigPath)
 	if err != nil {
 		return nil, err
 	}
 
-	targetCoreClient, err := kubernetes.NewForConfig(targetCoreKubeconfig)
-	if err != nil {
-		return nil, err
+	targetKubeconfig.Burst = *targetBurst
+	targetKubeconfig.QPS = float32(*targetQPS)
+	targetCoreClientBuilder := CoreControllerClientBuilder{
+		ClientConfig: targetKubeconfig,
 	}
+	targetCoreInformerFactory := coreinformers.NewSharedInformerFactory(
+		targetCoreClientBuilder.ClientOrDie("target-core-shared-informers"),
+		resyncPeriod,
+	)
+
+	coreSharedInformers := targetCoreInformerFactory.Core().V1()
+	nodeInformer := coreSharedInformers.Nodes().Informer()
 
-	manager := &McmManager{
+	m := &McmManager{
 		namespace:               namespace,
 		interrupt:               make(chan struct{}),
-		machineclient:           machineClient,
-		coreclient:              targetCoreClient,
+		machineClient:           controlMachineClient,
+		awsMachineClassLister:   machineSharedInformers.AWSMachineClasses().Lister(),
+		azureMachineClassLister: machineSharedInformers.AzureMachineClasses().Lister(),
+		machineClassLister:      machineSharedInformers.MachineClasses().Lister(),
+		machineLister:           machineSharedInformers.Machines().Lister(),
+		machineSetLister:        machineSharedInformers.MachineSets().Lister(),
 		machineDeploymentLister: machineSharedInformers.MachineDeployments().Lister(),
-		machinelisters:          machineSharedInformers.Machines().Lister(),
+		nodeLister:              coreSharedInformers.Nodes().Lister(),
 		discoveryOpts:           discoveryOpts,
 	}
 
-	return manager, nil
+	targetCoreInformerFactory.Start(m.interrupt)
+	controlMachineInformerFactory.Start(m.interrupt)
+
+	syncFuncs := []cache.InformerSynced{
+		awsMachineClassInformer.HasSynced,
+		azureMachineClassInformer.HasSynced,
+		machineClassInformer.HasSynced,
+		machineInformer.HasSynced,
+		machineSetInformer.HasSynced,
+		machineDeploymentInformer.HasSynced,
+		nodeInformer.HasSynced,
+	}
+	if !cache.WaitForCacheSync(m.interrupt, syncFuncs...) {
+		return nil, fmt.Errorf("Timed out waiting for caches to sync")
+	}
```

Comment:
Does this block indefinitely if you don't have all of the CRDs in your cluster?

Reply:
We might need a more explicit failure in that case, because previously cluster-autoscaler wouldn't complain. Probably it should log a fatal fault if any of the CRDs are missing.

Reply:
This is a good catch. I didn't think about this case. I have fixed it with this [here](https://github.com/gardener/autoscaler/pull/73/commits/3b73331aa56854d564ce4eba429dee02f58d8521#diff-fc87dde41e1a7abff0b0bc24dbabc258ac3d594a346b8f01e020b9e7f9775440R163-R207). I hope the changes on [this commit](https://github.com/gardener/autoscaler/pull/73/commits/3b73331aa56854d564ce4eba429dee02f58d8521) captures your concerns.